core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -820,131 +907,202 @@ class SocketServerTest extends JUnitSuite {
+    val allChannels = mutable.Set[String]()
+    val allLocallyClosedChannels = mutable.Set[String]()
+    val allDisconnectedChannels = mutable.Set[String]()
+    val allFailedChannels = mutable.Set[String]()
+
+    // Enable data from `Selector.poll()` to be deferred to a subsequent poll() until
+    // the number of elements of that type reaches `minPerPoll`. This enables tests to verify
+    // that failed processing doesn't impact subsequent processing within the same iteration.
+    class PollData[T] {
+      var minPerPoll = 1
+      val deferredValues = mutable.Buffer[T]()
+      val currentPollValues = mutable.Buffer[T]()
+      def update(newValues: mutable.Buffer[T]): Unit = {
+        if (!currentPollValues.isEmpty || deferredValues.size + newValues.size >= minPerPoll) {
+          if (!deferredValues.isEmpty) {
+            currentPollValues ++= deferredValues
+            deferredValues.clear()
+          }
+          currentPollValues ++= newValues
+        } else
+          deferredValues ++= newValues
+      }
+      def reset(): Unit = {
+        currentPollValues.clear()
+      }
+    }
+    val cachedCompletedReceives = new PollData[NetworkReceive]()
+    val cachedCompletedSends = new PollData[Send]()
+    val cachedDisconnected = new PollData[(String, ChannelState)]()
+    val allCachedPollData = Seq(cachedCompletedReceives, cachedCompletedSends, cachedDisconnected)
+    @volatile var minWakeupCount = 0
 
     def addFailure(operation: SelectorOperation, exception: Option[Exception] = None) {
-      val failureException = exception match {
-        case Some(t) => t
-        case None => new IllegalStateException(s"Test exception during $operation")
-      }
-      failures += operation -> failureException
+      failures += operation ->
+        exception.getOrElse(new IllegalStateException(s"Test exception during $operation"))
     }
 
-    private def onOperation(operation: SelectorOperation): Unit = {
+    private def onOperation(operation: SelectorOperation,
+        connectionId: Option[String] = None, undoAction: => Unit = {}): Unit = {
       operationCounts(operation) += 1
-      failures.get(operation).foreach(e => throw e)
+      failures.remove(operation).foreach { e =>
+        connectionId.foreach(allFailedChannels.add(_))
+        undoAction
+        throw e
+      }
     }
 
     def waitForOperations(operation: SelectorOperation, minExpectedTotal: Int): Unit = {
       TestUtils.waitUntilTrue(() =>
         operationCounts.getOrElse(operation, 0) >= minExpectedTotal, "Operations not performed within timeout")
     }
 
-    def runOp[T](operation: SelectorOperation)(code: => T): T = {
-      try {
-        code
-      } finally {
-        onOperation(operation)
-      }
+    def runOp[T](operation: SelectorOperation, connectionId: Option[String],
+        undoAction: => Unit = {})(code: => T): T = {
+      // If a failure is set on `operation`, throw that exception even if `code` fails
```

Comment:
Yes, that is a bit messy to do since the operations are performed on a different thread. Can be done by storing some state and checking later on, but not sure if it is worth the extra complexity.